source/vstd/bits.rs

```diff
@@ -0,0 +1,203 @@
+                    x as nat / pow2(shift as nat);
+                }
+            }
+        }
+
+        #[doc = "Proof that for all x and n of type "]
+        #[doc = stringify!($uN)]
+        #[doc = ", shifting x right by n is equivalent to division of x by 2^n."]
+        pub proof fn $name_auto()
+            ensures
+                forall|x: $uN, shift: $uN|
+                    0 <= shift < $uN::BITS ==> #[trigger] (x >> shift) == x as nat / pow2(shift as nat),
+        {
+            assert forall|x: $uN, shift: $uN| 0 <= shift < $uN::BITS implies #[trigger] (x >> shift) == x as nat
+                / pow2(shift as nat) by {
+                $name(x, shift);
+            }
+        }
+        }
+    };
+}
+
+lemma_shr_is_div!(lemma_u64_shr_is_div, lemma_u64_shr_is_div_auto, u64);
+lemma_shr_is_div!(lemma_u32_shr_is_div, lemma_u32_shr_is_div_auto, u32);
+lemma_shr_is_div!(lemma_u16_shr_is_div, lemma_u16_shr_is_div_auto, u16);
+lemma_shr_is_div!(lemma_u8_shr_is_div, lemma_u8_shr_is_div_auto, u8);
+
+// Proofs that a given power of 2 fits in an unsigned type.
+macro_rules! lemma_pow2_no_overflow {
+    ($name:ident, $name_auto:ident, $uN:ty) => {
+        verus! {
+        #[doc = "Proof that 2^n does not overflow "]
+        #[doc = stringify!($uN)]
+        #[doc = " for a given exponent n."]
+        pub proof fn $name(n: nat)
+            requires
+                0 <= n < $uN::BITS,
+            ensures
+                pow2(n) <= $uN::MAX,
+        {
+            lemma_pow2_strictly_increases(n, $uN::BITS as nat);
+            lemma2_to64();
+        }
+
+        #[doc = "Proof that 2^n does not overflow "]
+        #[doc = stringify!($uN)]
+        #[doc = " for all exponents in bounds."]
+        pub proof fn $name_auto()
+            ensures
+                forall|n: nat| 0 <= n < $uN::BITS ==> #[trigger] pow2(n) <= $uN::MAX,
+        {
+            assert forall|n: nat| 0 <= n < $uN::BITS implies #[trigger] pow2(n) <= $uN::MAX by {
+                $name(n);
+            }
+        }
+        }
+    };
+}
+
+lemma_pow2_no_overflow!(lemma_u64_pow2_no_overflow, lemma_u64_pow2_no_overflow_auto, u64);
+lemma_pow2_no_overflow!(lemma_u32_pow2_no_overflow, lemma_u32_pow2_no_overflow_auto, u32);
+lemma_pow2_no_overflow!(lemma_u16_pow2_no_overflow, lemma_u16_pow2_no_overflow_auto, u16);
+lemma_pow2_no_overflow!(lemma_u8_pow2_no_overflow, lemma_u8_pow2_no_overflow_auto, u8);
+
+// Proofs that shift left is equivalent to multiplication by power of 2.
+macro_rules! lemma_shl_is_mul {
+    ($name:ident, $name_auto:ident, $no_overflow:ident, $uN:ty) => {
+        verus! {
+        #[doc = "Proof that for given x and n of type "]
+        #[doc = stringify!($uN)]
+        #[doc = ", shifting x left by n is equivalent to multiplication of x by 2^n (provided no overflow)."]
+        pub proof fn $name(x: $uN, shift: $uN)
+            requires
+                0 <= shift < $uN::BITS,
+                x * pow2(shift as nat) <= $uN::MAX,
```

Comment:
Is this a reasonable requirement for the lemma? The alternative would be to prove that left shift equals `wrapping_mul`.